### PR TITLE
[WIP] Add bucketed join support for iceberg connector (#79)

### PR DESCRIPTION
**Pull request by hellium01**
## Description
This PR attempt to add bucketed join support to Iceberg connector. 

## Additional context and related issues

https://github.com/trinodb/trino/issues/19062

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
(x) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Adding bucketed join support for iceberg connector . ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by mosabua** [comment]
Please ensure that your build works locally and also here on github. Ping me if you need further help.

**Comment by mosabua** [comment]
@cwsteinbach fyi .. might be good to move this forward. 

@hellium01 would you be able to resolve conflicts?

Also fyi @raunaqmorarka

**Comment by mosabua** [comment]
@cla-bot check

**Comment by mosabua** [comment]
@hellium01 is seems your CLA is filed and processed - https://github.com/trinodb/cla/blob/9384d33f008b0835f9cff8e60880dde3bb237245/contributors#L18

but the clabot still fails .. this might have to do with your git and github config - the email in your commit message must be a configured email in your github profile. Ping me if you need help.

**Comment by wendigo** [comment]
@cla-bot check

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergNodePartitioningProvider.java [hunk 1]
```diff
@@ -14,30 +14,63 @@
...
     @Inject
-    public IcebergNodePartitioningProvider(TypeManager typeManager)
+    public IcebergNodePartitioningProvider(TypeManager typeManager, NodeManager nodeManager)
     {
         this.typeOperators = typeManager.getTypeOperators();
+        this.nodeManager = nodeManager;
+    }
+
+    @Override
+    public Optional<ConnectorBucketNodeMap> getBucketNodeMapping(
```
Comment (raunaqmorarka) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergNodePartitioningProvider.java:54]: Can you explain why implementing this function is needed ?
Is there a reason to care about which node the engine chooses to schedule a bucket on, as long as it schedules all splits for a bucket on the same node ?
The big comment below is picked up from hive connector code where we added this code to parallelize writes to hive tables with partitioning and bucketing. I'm not sure that's relevant to this PR.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergNodePartitioningProvider.java [hunk 2]
```diff
@@ -60,4 +102,17 @@ public BucketFunction getBucketFunction(
...
+    public ToIntFunction<ConnectorSplit> getSplitBucketFunction(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorPartitioningHandle partitioningHandle)
+    {
+        int maxBucketSize = nodeManager.getRequiredWorkerNodes().size() * PARTITIONED_BUCKETS_PER_NODE;
+        if (partitioningHandle instanceof CombinedBucketedPartitioningHandle || partitioningHandle instanceof BucketedTablePartitioningHandle) {
+            return split -> ((IcebergSplit) split).getBucketId()
+                    .orElseThrow(() -> new IllegalArgumentException("Bucket number not set in split")) % maxBucketSize;
+        }
+        return split -> {
+            throw new UnsupportedOperationException();
+        };
```
Comment (raunaqmorarka) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergNodePartitioningProvider.java:116]: Why not just
```
return value -> ((IcebergSplit) value). getBucketId()
                .orElseThrow(() -> new IllegalArgumentException("Bucket number not set in split"));
```

**Review by raunaqmorarka — COMMENTED** [review]
Some preliminary comments

**Comment by raunaqmorarka** [comment]
Most of this is superseded by https://github.com/trinodb/trino/pull/23432
There is potential for follow-up work, but it should be based on top of that PR.

**Pull request closed without merging**